Add trait_upcasting related languages changes (#1622)
rust-lang/reference: src/type-coercions.md

* `T` to `dyn U`, when `T` implements `U + Sized`, and `U` is [object safe].

* `dyn T` to `dyn U` when `U` is one of `T`'s [supertraits].

>     * This allows dropping auto traits, i.e. `dyn T + Auto` to `dyn U` is allowed.

Reviewer: This also allows adding auto traits, i.e. `trait Sub: Sup + Auto` you can do `dyn Sub` -> `dyn Sup + Auto`. https://github.com/rust-lang/rust/pull/119338

Author: Added a mention of that

r[coerce.unsize.intro]
The following coercions are called `unsized coercions`, since they

> relate to converting types to unsized types, and are permitted in a few

Reviewer 2: ```suggestion
relate to converting sized types to unsized types or converting from one unsized type to another, and are permitted in a few
```

(more clarity, same meaning)

Author: For me your proposed change kinda flushes the cache, I had to re-read it 4 times before I figured out what it means 💀 (then again, maybe I shouldn't be reading it at 3am)

Since it's the same meaning I'd think that more concise is better? Do you disagree?

Author: @traviscross you've applied the suggestion without addressing my concern, so I reverted the commit (5dbb2de1f49758d50eeae21865fe75148c88b15f). I don't think this change improves this paragraph -- it changes a general statements into a union of concrete, which just makes it harder to read. Moreover, my refactor of the unsizing section also changes this paragraph anyway, so I'd wait for the refactor PR to be open before trying to fix anything.

Reviewer 3: I had talked with Niko awhile back, and he had wanted to keep it, and it seems clearer to me as well.  What if were to just say both, e.g.?:

```suggestion
relate to converting types to unsized types (that is, converting sized types to unsized types or converting from one unsized type to another), and are permitted in a few
```

Author: That might be fine, but again, my refactor already contains similar explanations, so I'd suggest discussing this once it's open.

Reviewer 3: OK.  We'll handle this is your later PR.

cases where other coercions are not, as described above. They can still happen
anywhere else a coercion can occur.

r[coerce.unsize.misnomer]

> > Note: "unsizing" is a bit of a misnomer,
> > since this covers unsized->unsized coercions too.

Reviewer 3: ```suggestion
> Note: These are sometimes called "unsizing" coercions, but that can be a bit of a misnomer, since these cover unsized->unsized coercions too.
```

Not sure if something like this is what you meant to say or not.  The Reference doesn't refer to "unsizing" anywhere -- even though, you're right, we often do in conversation -- so we'd need to introduce that somehow in this note.

Author: Hm. There may not be the literal word "unsizing", but we do refer to this coercion as "unsize coercion". 

> The following coercions are called `unsized coercions`, since they

I do feel like it's clear what this means, given that this note is surrounded by mentions of unsize coercions. (but maybe not see you seem to be confused?)

Author: Although I'd also say that replacing "unsized coercion" to "unsizing coercion" everywhere in the reference would be a good idea. It sounds better and more closely matches how people actually talk about this feature.

Reviewer 4: In conversation with @traviscross, I was one of the confused ones. TC pointed out that "unsized coercion" can be interpreted as the codomain so to speak. That is, it doesn't seem like a misnomer to me because it is only referring to coercions *to* something not sized, and thus seemed accurate.

It is true the word "unsizing" implies "to remove the size of", that is the *domain* is sized, and thus is not always an accurate term. However, we don't use "unsizing" here.

Probably a bit of a muddled interpretation on my part, but does that make sense?

Reviewer 5: "unsized coercion" sounds like the coercion itself is unsized. I don't even know what that should mean.

Reviewer 3: I'm just removing this note section for now.  We can handle this in later work.

r[coerce.unsize.trait-object]
* `T` to `dyn U`, when `T` implements `U + Sized`, and `U` is [dyn compatible].

> r[coerce.unsize.trait-upcast]

Author: I think this whole section might need a retractor :p

Specifically I think it should more clearly state the difference between `Unsize` and `CoerceUnsize`, mention that the former is not actually a coercion, and the latter is only for pointers. 

Then it would also make sense to add a note that unsizing coercion changes the metadata of a pointer, while all other coercions/casts don't.

But, if you don't mind I'll leave this as a follow up ^^'

> r[coerce.unsize.trait-upcast]

Reviewer 3: OK, we'll leave this for the follow-up PR you have in progress.

* `dyn T` to `dyn U`, when `U` is one of `T`'s supertraits.

> * `dyn T` to `dyn U` when `U` is one of `T`'s [supertraits].

Reviewer 3: These lines seem duplicates.
